File: src/FSharpVSPowerTools.Pure/Commands/StandardCommandDispatcher.fs
namespace FSharpVSPowerTools

open System
open System.Diagnostics
open Fantomas
open FSharpVSPowerTools.CodeFormatting
open Microsoft.VisualStudio
open Microsoft.VisualStudio.Shell
open Microsoft.VisualStudio.OLE.Interop
open Microsoft.VisualStudio.Text.Editor
open Microsoft.VisualStudio.TextManager.Interop


type private CommandMapping (commandGroup:Guid, commandId:int, commandOptions:OLECMDF, commandCreator:unit->CommandBase) =

    member __.Matches (otherCommandGroup:Guid, otherCommandId:int) =
            commandGroup = otherCommandGroup && commandId = otherCommandId
    
    member __.GetMatchingIndex (oCommandGroup:Guid,commands:OLECMD []) =
        if commandGroup = oCommandGroup then
            commands |> Array.findIndex (fun cmd-> cmd.cmdID = uint32 commandId)
        else -1

    member __.CommandOptions with get () = uint32 commandOptions
   
    member __.CreateCommand () = commandCreator ()


type StandardCommandDispatcher private () as self =

    [<DefaultValue>] val mutable private textView       : IWpfTextView
    [<DefaultValue>] val mutable private codeServices   : CodeFormattingServices



//    let mutable codeServices    = Unchecked.defaultof<CodeFormattingServices>
    let mutable commandChain    = Unchecked.defaultof<IOleCommandTarget>
    let mutable commands        = [||]


    member __.GetConfig () =
        let editorOptions = self.codeServices.EditorOptionsFactory.GetOptions self.textView.TextBuffer
        let indentSize    = editorOptions.GetOptionValue (IndentSize ()).Key
        let customOptions = Setting.getFormattingOptions self.codeServices.ServiceProvider

        FormatConfig.FormatConfig.create
            (   indentSize                              ,
                customOptions.PageWidth                 ,
                customOptions.SemicolonAtEndOfLine      ,
                customOptions.SpaceBeforeArgument       ,
                customOptions.SpaceBeforeColon          ,
                customOptions.SpaceAfterComma           ,
                customOptions.SpaceAfterSemicolon       ,
                customOptions.IndentOnTryWith           ,
                customOptions.ReorderOpenDeclaration    ,
                customOptions.SpaceAroundDelimiter      )


    member private __.StandardCommandDispatcher () =
        commands <-
            [|  CommandMapping
                    (   typeof<VSConstants.VSStd2KCmdID>.GUID,
                        int VSConstants.VSStd2KCmdID.FORMATDOCUMENT,
                        OLECMDF.OLECMDF_ENABLED ||| OLECMDF.OLECMDF_SUPPORTED,
                        fun () -> FormatDocumentCommand (Func<_>(self.GetConfig)) :> CommandBase )
                CommandMapping
                    (   typeof<VSConstants.VSStd2KCmdID>.GUID,
                        int VSConstants.VSStd2KCmdID.FORMATSELECTION,
                        OLECMDF.OLECMDF_ENABLED ||| OLECMDF.OLECMDF_SUPPORTED,
                        fun () -> FormatDocumentCommand (Func<_>(self.GetConfig)) :> CommandBase )
            |]

    // this probably shouldn't be an option
    static member Register (interopTextView :IVsTextView,  
                            textView        :IWpfTextView, 
                            codeServices    :CodeFormattingServices ) : StandardCommandDispatcher option =
        //let generalOptions = Setting.getGeneralOptions codeServices.ServiceProvider
        //Unchecked.defaultof<StandardCommandDispatcher>
        maybe{
         //   let!  generalOptions = Setting.tryGetGeneralOptions codeServices.ServiceProvider
 //               return!
           //     if not generalOptions.FormattingEnabled then None else
                let dispatcher = StandardCommandDispatcher()
                dispatcher.textView <- textView
                return!
                    Some dispatcher
        }
 

        
    interface IOleCommandTarget with

        member __.QueryStatus (pguidCmdGroup: byref<Guid>, cCmds: uint32, prgCmds: OLECMD [], pCmdText: nativeint): int = 
            for commandMapping in commands do
                let fmtindex = commandMapping.GetMatchingIndex (pguidCmdGroup, prgCmds)
                if  fmtindex >= 0 then
                    prgCmds.[fmtindex].cmdf <- commandMapping.CommandOptions
                    VSConstants.S_OK 
                    |> ignore // THIS PART IS PROBABLY VERY WRONG
            commandChain.QueryStatus(ref pguidCmdGroup, cCmds, prgCmds, pCmdText)           


        member __.Exec (pguidCmdGroup: byref<Guid>, nCmdID: uint32, nCmdexecopt: uint32, pvaIn: nativeint, pvaOut: nativeint): int = 
            for commandMapping in commands do
                let command = commandMapping.CreateCommand()
                command.TextView <- self.textView
                command.Services <- self.codeServices
                command.Execute()
                VSConstants.S_OK 
                |> ignore // THIS PART IS PROBABLY VERY WRONG
            commandChain.Exec(ref  pguidCmdGroup, nCmdID, nCmdexecopt, pvaIn, pvaOut)           

File: src/FSharpVSPowerTools/Commands/StandardCommandDispatcher.cs
﻿using System;
using System.Diagnostics;
using Fantomas;
using FSharpVSPowerTools.CodeFormatting;
using Microsoft.VisualStudio;
using Microsoft.VisualStudio.Shell;
using Microsoft.VisualStudio.OLE.Interop;
using Microsoft.VisualStudio.Text.Editor;
using Microsoft.VisualStudio.TextManager.Interop;
using System.ComponentModel.Composition;

namespace FSharpVSPowerTools
{
    public class StandardCommandDispatcher : IOleCommandTarget
    {
        private IWpfTextView _textView;
        private CodeFormattingServices _services;
        private IOleCommandTarget _commandChain;
        private readonly CommandMapping[] Commands;

        [Import]
        private static  IGeneralOptions generalOptions { get; set; }
        [Import]
        private static IFormattingOptions formattingOptions { get; set; }




        public static StandardCommandDispatcher Register(IVsTextView interopTextView, IWpfTextView textView, CodeFormattingServices services)
        {
            //var generalOptions = Setting.getGeneralOptions(services.ServiceProvider);
            if (generalOptions == null || !generalOptions.FormattingEnabled) return null;

            var dispatcher = new StandardCommandDispatcher();
            dispatcher._textView = textView;
            dispatcher._services = services;
            interopTextView.AddCommandFilter(dispatcher, out dispatcher._commandChain);
            return dispatcher;
        }

        private StandardCommandDispatcher()
        {
            Commands = new[]
            {
                new CommandMapping(
                    typeof (VSConstants.VSStd2KCmdID).GUID,
                    (int) (VSConstants.VSStd2KCmdID.FORMATDOCUMENT),
                    OLECMDF.OLECMDF_ENABLED | OLECMDF.OLECMDF_SUPPORTED,
                    () => new FormatDocumentCommand(this.GetConfig)),
                new CommandMapping(
                    typeof (VSConstants.VSStd2KCmdID).GUID,
                    (int) (VSConstants.VSStd2KCmdID.FORMATSELECTION),
                    OLECMDF.OLECMDF_ENABLED | OLECMDF.OLECMDF_SUPPORTED,
                    () => new FormatSelectionCommand(this.GetConfig))
            };
        }

        private FormatConfig.FormatConfig GetConfig()
        {
            IEditorOptions editorOptions = _services.EditorOptionsFactory.GetOptions(_textView.TextBuffer);
            int indentSize = editorOptions.GetOptionValue((new IndentSize()).Key);
            //var customOptions = Setting.getFormattingOptions(_services.ServiceProvider);
            var customOptions = formattingOptions;

            var config =
                FormatConfig.FormatConfig.create(
                    indentSize,
                    customOptions.PageWidth,
                    customOptions.SemicolonAtEndOfLine,
                    customOptions.SpaceBeforeArgument,
                    customOptions.SpaceBeforeColon,
                    customOptions.SpaceAfterComma,
                    customOptions.SpaceAfterSemicolon,
                    customOptions.IndentOnTryWith,
                    customOptions.ReorderOpenDeclaration,
                    customOptions.SpaceAroundDelimiter);

            return config;
        }

        public int QueryStatus(ref Guid pguidCmdGroup, uint cCmds, OLECMD[] prgCmds, IntPtr pCmdText)
        {
            foreach (var commandMapping in Commands)
            {
                int fmtindex = commandMapping.GetMatchingIndex(pguidCmdGroup, prgCmds);
                if (fmtindex >= 0)
                {
                    prgCmds[fmtindex].cmdf = commandMapping.CommandOptions;
                    return VSConstants.S_OK;
                }
            }

            return _commandChain.QueryStatus(pguidCmdGroup, cCmds, prgCmds, pCmdText);
        }

        public int Exec(ref Guid pguidCmdGroup, uint nCmdID, uint nCmdexecopt, IntPtr pvaIn, IntPtr pvaOut)
        {
            foreach (var commandMapping in Commands)
            {
                if (commandMapping.Matches(pguidCmdGroup, nCmdID))
                {
                    var command = commandMapping.CreateCommand();
                    command.TextView = _textView;
                    command.Services = _services;
                    command.Execute();
                    return VSConstants.S_OK;
                }
            }

            return _commandChain.Exec(pguidCmdGroup, nCmdID, nCmdexecopt, pvaIn, pvaOut);
        }

        private class CommandMapping
        {
            private readonly Guid _commandGroup;
            private readonly uint _commandId;
            private readonly uint _commandOptions;
            private readonly Func<CommandBase> _commandCreator;

            public CommandMapping(Guid commandGroup, int commandId, OLECMDF commandOptions, Func<CommandBase> commandCreator)
            {
                _commandGroup = commandGroup;
                _commandId = (uint)commandId;
                _commandOptions = (uint)commandOptions;
                _commandCreator = commandCreator;
            }

            public bool Matches(Guid commandGroup, uint commandId)
            {
                return commandGroup == _commandGroup && commandId == _commandId;
            }

            public int GetMatchingIndex(Guid commandGroup, OLECMD[] commands)
            {
                if (commandGroup == _commandGroup)
                {
                    return System.Array.FindIndex(commands, cmd => cmd.cmdID == _commandId);
                }
                return -1;
            }

            public uint CommandOptions
            {
                get { return _commandOptions; }
            }

            public CommandBase CreateCommand()
            {
                return _commandCreator();
            }
        }
    }
}
